src/libs/highlevel/elektra.c

```diff
@@ -0,0 +1,233 @@
+/**
+ * @file
+ *
+ * @brief Elektra High Level API.
+ *
+ * @copyright BSD License (see doc/LICENSE.md or http://www.libelektra.org)
+ */
+
+#include "elektra.h"
+#include "elektra_conversion.h"
+#include "elektra_error_private.h"
+#include "elektra_private.h"
+#include "elektra_types.h"
+#include "kdbhelper.h"
+#include "kdblogger.h"
+#include <stdlib.h>
+
+static void defaultFatalErrorHandler (ElektraError * error)
+{
+	ELEKTRA_LOG_DEBUG ("FATAL ERROR [%d]: %s", error->code, error->description);
+	ElektraErrorCode code = error->code;
+	elektraFree (error);
+	exit (code);
+}
+
+/**
+ * \defgroup highlevel High-level API
+ * @{
+ */
+
+ELEKTRA_TAG_DEFINITIONS (const char *, String, KDB_TYPE_STRING, elektraStrDup, elektraKeyToString)
+ELEKTRA_TAG_DEFINITIONS (kdb_boolean_t, Boolean, KDB_TYPE_BOOLEAN, elektraBooleanToString, elektraKeyToBoolean)
+ELEKTRA_TAG_DEFINITIONS (kdb_char_t, Char, KDB_TYPE_CHAR, elektraCharToString, elektraKeyToChar)
+ELEKTRA_TAG_DEFINITIONS (kdb_octet_t, Octet, KDB_TYPE_OCTET, elektraOctetToString, elektraKeyToOctet)
+ELEKTRA_TAG_DEFINITIONS (kdb_short_t, Short, KDB_TYPE_SHORT, elektraShortToString, elektraKeyToShort)
+ELEKTRA_TAG_DEFINITIONS (kdb_unsigned_short_t, UnsignedShort, KDB_TYPE_UNSIGNED_SHORT, elektraUnsignedShortToString,
+			 elektraKeyToUnsignedShort)
+ELEKTRA_TAG_DEFINITIONS (kdb_long_t, Long, KDB_TYPE_LONG, elektraLongToString, elektraKeyToLong)
+ELEKTRA_TAG_DEFINITIONS (kdb_unsigned_long_t, UnsignedLong, KDB_TYPE_UNSIGNED_LONG, elektraUnsignedLongToString, elektraKeyToUnsignedLong)
+ELEKTRA_TAG_DEFINITIONS (kdb_long_long_t, LongLong, KDB_TYPE_LONG_LONG, elektraLongLongToString, elektraKeyToLongLong)
+ELEKTRA_TAG_DEFINITIONS (kdb_unsigned_long_long_t, UnsignedLongLong, KDB_TYPE_UNSIGNED_LONG_LONG, elektraLongLongToString,
+			 elektraKeyToUnsignedLongLong)
+ELEKTRA_TAG_DEFINITIONS (kdb_float_t, Float, KDB_TYPE_FLOAT, elektraFloatToString, elektraKeyToFloat)
+ELEKTRA_TAG_DEFINITIONS (kdb_double_t, Double, KDB_TYPE_DOUBLE, elektraDoubleToString, elektraKeyToDouble)
+
+#if defined(HAVE_SIZEOF_LONG_DOUBLE) && (SIZEOF_LONG_DOUBLE == 16 || SIZEOF_LONG_DOUBLE == 12)
+
+ELEKTRA_TAG_DEFINITIONS (kdb_long_double_t, LongDouble, KDB_TYPE_LONG_DOUBLE, elektraLongDoubleToString, elektraKeyToLongDouble)
+
+#endif // HAVE_SIZEOF_LONG_DOUBLE
+
+ELEKTRA_TAG_DEFINITIONS (int, Enum, KDB_TYPE_ENUM, elektraLongToString, elektraKeyToLong)
+
+/**
+ * Initializes a new Elektra instance.
+ * @param application The parent key for your application.
```

Comment:
> Btw. what about also accepting relative keys which will build up the string `"/sw/org/<what the user passed>/#0/current"`

That seems very limited. Shouldn't the application use its own `orgname` instead of `org`, and the `current` part should ideally be set by the user of the application not the application itself to support multiple profiles.

The better so solution, in my mind, would be to provide a separate method that constructs the whole string with the signature: `char* foo(const char* org, const char* app, int version, const char* profile)`

Reply:
I mean providing both: relative paths are `"/sw/org/<what the user passed>/#0/current"` but absolute are still accepted. But maybe that is a useless convenience feature because sooner or later one needs to know the path anyway.

Having a signature that provides all parts separately is of course more convenient from C-programming point of view but it is also less flexible (e.g. if someone needs /hw instead of /sw).

Profile actually should not be part of the specification because by convention always "current" should be used.

So maybe it is best to leave it as-is.

Reply:
> Having a signature that provides all parts separately is of course more convenient from C-programming point of view but it is also less flexible (e.g. if someone needs /hw instead of /sw).

Well the first part could also be provided to the function. 
`char* foo(const char * base, const char * org, const char * app, int version)`
would basically be
`elektraFormat("/%s/%s/%s/#%d/current", base, org, app, version);`

For me the problem would actually be more finding a good name for such a function...

Reply:
Yes, the name of the function is the one problem and that it has too many parameters is the other. And finally you need the full name so that it will work with `kdb set`. So better to leave it as is.